[RDY] unittest: Allow multiple indirect includes

ZyX-I opened the pull request:
Works by saving all preprocessor defines and reusing them on each run. This also 
saves NVIM_HEADER_H defines. Saving other defines is needed for defines like 
`Map(foo, bar)` which are sometimes used to declare types or functions. Saving 
types or function declarations is not needed because they are recorded as luajit 
state.

Fixes #5857

ZyX-I committed:
unittest: Allow multiple indirect includes

Works by saving all preprocessor defines and reusing them on each run. This also
saves NVIM_HEADER_H defines. Saving other defines is needed for defines like
`Map(foo, bar)` which are sometimes used to declare types or functions. Saving
types or function declarations is not needed because they are recorded as luajit
state.

Fixes #5857

ZyX-I committed:
unittest: Record previous defines in another place

Previous commit made preprocess.lua know how its output will be used. This moves 
state to cimport, making only it know which is cleaner.

justinmk commented on test/unit/helpers.lua:
```diff
@@ -68,14 +68,12 @@ local function cimport(...)
 
   if body == nil then
-    print("ERROR: helpers.lua: Preprocess.preprocess_stream():read() returned empty")
+    print("ERROR: helpers.lua: Preprocess.preprocess() returned empty")
```
This sanity check was for the `:read()` call. Because of a lua or lfs bug it sometimes returns empty, and catching it here saves debugging later. Not sure if it's still needed, but if so it so check `ret == nil` after line 182 in `test/unit/preprocess.lua`.

justinmk commented on test/unit/preprocess.lua:
```diff
@@ -145,21 +149,40 @@ end
                tostring(defines) ..
-               " -std=c99 -P -E -"):gsub('$hdr', shell_quote(pseudoheader))
+               " -std=c99 -P -E " .. shell_quote(pseudoheader_fname))
+  local def_cmd = ("echo $hdr | " ..
```
Does `$hdr` get replaced now? Before `:gsub('$hdr', ...)` did so.

ZyX-I replied:
`echo $hdr | … -` was replaced with actual file name. Just forgot to remove the command.

ZyX-I replied:
Currently it works because any input is ignored.

justinmk commented on test/unit/preprocess.lua:
```diff
@@ -145,21 +149,40 @@ end
+  pseudoheader_file:flush()
+  pseudoheader_file:close()
   local defines = table.concat(self.preprocessor_extra_flags, ' ')
   local cmd = ("echo $hdr | " ..
```
Would be slightly easier to follow if this were defined next to `io.popen(cmd)` below.

ZyX-I committed:
unittest: Move nil checks to Gcc:preprocess

ZyX-I committed:
unittest: Refactor preprocess.lua

Keeps arguments separated and not joined as a single string as long as possible. 
Abstracts away additional arguments so that Gcc:preprocess should work for 
compilers with different conventions should they be supported.

justinmk commented on test/unit/preprocess.lua:
```diff
@@ -61,12 +65,12 @@ end
 -- #include <math.h>
 --
--- headerize({"vim.h", "memory.h", false}
+-- headerize({"vim.h", "memory.h"}, false}
```
Should the final `}` be a `)` ?

ZyX-I committed:
unittest: Fix linter errors

ZyX-I commented:
This is also ready, quickbuild failure seems to be one of the unrelated random failures:

```
07:48:57,849 INFO  - not ok 107 - TermClose event works as expected
07:48:57,849 INFO  - # test/functional/autocmd/termclose_spec.lua @ 20
07:48:57,849 INFO  - # Failure message: ./test/functional/ui/screen.lua:318: Row 4 did not match.
07:48:57,849 INFO  - # Expected:
07:48:57,849 INFO  - # |ready $ |
07:48:57,849 INFO  - # |[Process exited 0] |
07:48:57,849 INFO  - # |^ |
07:48:57,849 INFO  - # |*TermClose works! |
07:48:57,849 INFO  - # Actual:
07:48:57,849 INFO  - # |ready $ |
07:48:58,079 INFO  - # |[Process exited 0] |
07:48:58,080 INFO  - # |^ |
07:48:58,080 INFO  - # |* |
07:48:58,080 INFO  - #
07:48:58,080 INFO  - # To print the expect() call that would assert the current screen state, use
07:48:58,080 INFO  - # screen:snaphot_util(). In case of non-deterministic failures, use
07:48:58,080 INFO  - # screen:redraw_debug() to show all intermediate screen states.
07:48:58,080 INFO  - # stack traceback:
07:48:58,080 INFO  - #     ./test/functional/ui/screen.lua:318: in function 'wait'
07:48:58,080 INFO  - #     ./test/functional/ui/screen.lua:229: in function 'expect'
07:48:58,080 INFO  - #     test/functional/autocmd/termclose_spec.lua:24: in function <test/functional/autocmd/termclose_spec.lua:20>
07:48:58,080 INFO  - #
07:48:58,080 INFO  - # test/functional/autocmd/termclose_spec.lua @ 32: TermClose event reports the correct <abuf>
```
